Make seastar options HUP-updateable

xemul opened the pull request:
Currently all seastar options are immutable after seastar app starts. That's not very convenient as it takes full app restart to pick up new options if needed. Some options, however, allow live-updating because they either just one-to-one map to some simple reactor variable or because live-update can be implemented as a callable.

Seastar can read its options from ini-style seastar.conf file, so what this PR does is makes seastar re-read this file upon receiving HUP signal (*). Options should be marked as live-updateable explicitly by providing a noncopyable function on construction. When config file is re-read all those functions are called with new values arguments.

As an example the --unsafe-bypass-fsync is turned into live-updateable option.

(*) Scylla re-installs HUP handler, so it'll have to call reread_config() by hand.

dorlaor commented:
Finally we have it! Thanks for caring

On Fri, Apr 14, 2023 at 10:25 AM Pavel Emelyanov ***@***.***>
wrote:

> Currently all seastar options are immutable after seastar app starts.
> That's not very convenient as it takes full app restart to pick up new
> options if needed. Some options, however, allow live-updating because they
> either just one-to-one map to some simple reactor variable or because
> live-update can be implemented as a callable.
>
> Seastar can read its options from ini-style seastar.conf file, so what
> this PR does is makes seastar re-read this file upon receiving HUP signal
> (*). Options should be marked as live-updateable explicitly by providing a
> noncopyable function on construction. When config file is re-read all those
> functions are called with new values arguments.
>
> As an example the --unsafe-bypass-fsync is turned into live-updateable
> option.
>
> (*) Scylla re-installs HUP handler, so it'll have to call reread_config()
> by hand.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scylladb/seastar/pull/1615
> Commit Summary
>
>    - 2994fee
>    <https://github.com/scylladb/seastar/pull/1615/commits/2994fee38b8527a2fa1406ad2673f0cf4e9d3e53>
>    app-template: Add --config-file option
>    - dff39c8
>    <https://github.com/scylladb/seastar/pull/1615/commits/dff39c8470e273d1032d658fb0f334e693b96f3f>
>    app-template,program-options: Kick config file re-read on HUP
>    - be588f4
>    <https://github.com/scylladb/seastar/pull/1615/commits/be588f4860b9b0421521cf7f48871c187e68081e>
>    program_options: Introduce updateable_value<>
>    - 8780210
>    <https://github.com/scylladb/seastar/pull/1615/commits/8780210ef85efc8dc051067208b3480eb3220230>
>    reactor: Make unsafe-bypass-fsync updateable_value
>
> File Changes
>
> (6 files <https://github.com/scylladb/seastar/pull/1615/files>)
>
>    - *M* include/seastar/core/app-template.hh
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-3e190d644dc87f6078b21e63c0a9810ed11149b254fac0ee0835e6af78fab8c7>
>    (3)
>    - *M* include/seastar/core/reactor_config.hh
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-fd0fd4159888fc76a61154ea3374f2dc4324f1a87f2bc5bee2a99378375ec7ce>
>    (2)
>    - *M* include/seastar/util/program-options.hh
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-caeaa5bcacc9586a4a16ff15a03b27c68be1ebb7b9819aefb18f98213e8c6a68>
>    (31)
>    - *M* src/core/app-template.cc
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-e2dd21fb36d34d21fc9aa54ecf80b6251c831a79d22572db2b9451e06f7569af>
>    (22)
>    - *M* src/core/reactor.cc
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-5e5dfd46a396d71e0d68f183d6eb0f3bb5aad37be914323b8e4fea134dafec0c>
>    (6)
>    - *M* src/util/program-options.cc
>    <https://github.com/scylladb/seastar/pull/1615/files#diff-6f4535603abb308fbf9f665e3b54f5ca9b48a77fdb84ed0b87cb29a7c9e644c5>
>    (18)
>
> Patch Links:
>
>    - https://github.com/scylladb/seastar/pull/1615.patch
>    - https://github.com/scylladb/seastar/pull/1615.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/seastar/pull/1615>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHUROXFHTEP3A5EB5VC5DXBGCBRANCNFSM6AAAAAAW6V7Z7Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

denesb commented on src/core/app-template.cc:
```diff
@@ -56,6 +56,7 @@ seastar_options_from_config(app_template::config cfg) {
 
 app_template::seastar_options::seastar_options()
     : program_options::option_group(nullptr, "seastar")
+    , config_file(*this, "config-file", std::nullopt, "Path to seastar.conf file")
```
We can put here the current default (`~/.config/seastar/seastar.conf`), so it appears in command line help and we also avoid the special case below.
The slight drawback is that we have to implement substituting `~`, but that is simple enough.

xemul replied:
Is there std::/boost:: tilde expansion? [It's not just](https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html) `s/~/${HOME}/` and I don't feel enough courage to implement it thoroughly.

denesb replied:
I don't know of any such facilities, and I don't think its worth it just for this. But please do mention in the option description that we do have a default.

tchaikov replied:
why do we need to implement the tilde expansion? 

we already have `std::string(home) + "/.config/seastar/seastar.conf"`. i think what is missing is to extract it into a static helper function, which returns an `optional<std::string>`, so the ctor of `seastar_options` can use it for the default value of `config_file`.

denesb replied:
Placing expanded home path into default value, makes the help text context dependant. A `--help` ran as `user1` will have incorrect value for `user2`.

tchaikov replied:
i see. we need to expand it when actually consuming it.

denesb replied:
Yes.

denesb commented on include/seastar/util/program-options.hh:
```diff
@@ -463,6 +466,27 @@ public:
+    noncopyable_function<void(const updateable_value&)> _update;
+    virtual void do_mutate_live(options_mutator& mutator) override {
+        value<T>::do_mutate(mutator);
+        _update(*this);
```
This will unconditionally call `_update()` for all values, regardless of whether their value changed or not. Not sure if this is a problem.
We could make `do_mutate()` return `bool` and return `true` if value changed. We don't need to try very hard to detect changes, it is enough if we detect those cases where a value wasn't updated at all (it is still default). Most options are left as default anyway and are never updated.

xemul replied:
Good point. I like the latter approach so the plain

```c++
    virtual void do_mutate_live(options_mutator& mutator) override {
        value<T>::do_mutate(mutator);
        if (!_defaulted) {
            _update(*this);
        }
       ...
```

should work well enough

denesb replied:
Note that we have to detect the case where an option that previously had a value, is now defaulted (option was removed from settings file altogether). This won't detect that.
Actually, I think the current code will not even update such an option at all as `mutator.visit_value(val)` will return false for such values. When doing a live update and the visitor returns false from `visit_value` (meaning it has no value for this setting), we should reset the option back to default.

So we would invoke the updater in two cases: 
* `visit_value()` returned `true`
* `visit_value()` returned `false` and `_defaulted` was `false` (the value is being reset to default).

tchaikov commented on src/core/app-template.cc:
```diff
@@ -107,7 +108,12 @@ const app_template::seastar_options& app_template::options() const {
-        if (home) {
+        if (configuration.contains("config-file")) {
+            std::ifstream ifs_io(configuration["config-file"].as<std::string>());
+            if (ifs_io) {
```
i think, if user explicitly specifies a `config-file`, but that file is not readable, probably we should fail the whole application instead of swallow the failure and move on. so the user is aware that the option is invalid or something very bad happened.

tchaikov commented on src/core/app-template.cc:
```diff
@@ -107,7 +108,12 @@ const app_template::seastar_options& app_template::options() const {
+            if (ifs_io) {
+                bpo::store(bpo::parse_config_file(ifs_io, _opts_conf_file), configuration);
+            }
+        } else if (home) {
```
nit,
```suggestion
        } else if (auto home = std::getenv("HOME"); home) {
```

xemul replied:
Good nit, but no longer relevant

tchaikov commented on src/core/app-template.cc:
```diff
@@ -126,6 +126,16 @@ app_template::configuration_reader app_template::get_default_configuration_reade
 
+void app_template::reread_config() {
+    bpo::variables_map configuration;
+    if (_configuration->contains("config-file")) {
+        configuration.emplace("config-file", (*_configuration)["config-file"]);
+    }
+    _conf_reader(configuration);
+    program_options::variables_map_extracting_visitor visitor(configuration);
+    _opts.mutate_live(visitor);
```
i am wondering if we can just skip the reread if `!_configuration->contains("config-file")`? as there is no point of reading a conffile unless the file is specified in the command line or in the previous config file. so might want to put something like:
```c++
auto config_file = _configuration->find("config-file");
if (config_file == _configuration->end()) {
    return;
}
// reread on demand
```

please note, in order to avoid double-lookup, i am using `find()` here. not a big deal, but it also helps remove the one more repeating of "`config-file`".

xemul replied:
> i am wondering if we can just skip the reread if !_configuration->contains("config-file")? as there is no point of reading a conffile unless the file is specified in the command line or in the previous config file

HUP may still mean that the default $HOME/.config/seastar/seastar.conf was updated and needs to be re-read.

tchaikov replied:
ahh, right!

xemul commented:
upd:
- make reread_config() public so that app could kick it by hand
- call updating callback when the option is "back to default"
- protect against accidental HUP re-entrance
- default --config-file to be ~/.config/seastar/seastar.conf
- fail start in case --config-file is specified, but the file is not readable for whatever reason

denesb commented on include/seastar/util/program-options.hh:
```diff
@@ -463,6 +466,61 @@ public:
+            value<T>::do_set_value(std::move(val), false);
+            _update(*this, lock(*this));
+        } else if (!value<T>::defaulted()) {
+            value<T>::do_set_value(*_default_value, true);
```
This will fail if `_default_value` is `nullopt`. You should add a `do_set_value()` overload which takes an optional, so you can pass `_default_value` as-is.

xemul replied:
Yikes, indeed

denesb reviewed (commented):
LGTM, one minor correctness issue.

xemul commented:
upd:
- handle restore-default case for nullopt defaults

tchaikov commented on src/core/app-template.cc:
```diff
@@ -107,11 +108,19 @@ const app_template::seastar_options& app_template::options() const {
         auto home = std::getenv("HOME");
+
+        std::string config_file = _opts.config_file.get_value();
+        if (config_file[0] == '~' && home) {
```
sorry for being paranoid, but better off checking for empty `config_file`. and add fail early if `$HOME` is not available. the latter is much less likely to happen, but still..
like:
```c++
// use starts_with() in C++20
if (config_file.compare(0, 1, "~") == 0) {
    auto home = std::getenv("HOME");
    if (!home) {
        throw std::runtime_error(format("Cannot expand 'config-file': {}", config_file));
    }
    config_file.replace(0, 1, home);
}
```

i made couple changes:

1. use `compare()` instead of assuming `!config_file.empty()`. explained above already.
2.  throws if `$HOME` is not set, this is a fatal error. although it's quite like the "Cannot read config file" error, but a different one.
3. use `home` instead of `std::string(home)`, because of the 5th usage in https://en.cppreference.com/w/cpp/string/basic_string/replace

xemul replied:
Yup, good suggestion. Will fix shortly

tchaikov commented on src/core/app-template.cc:
```diff
@@ -191,10 +200,15 @@ app_template::run_deprecated(int ac, char ** av, std::function<void ()>&& func)
         fmt::print("error: {}\n\nTry --help.\n", e.what());
         return 2;
+    } catch (...) {
+        fmt::print("error: {}\n\n", std::current_exception());
```
better off sending error messages to stderr, like
```c++
fmt::print(stderr, "error: {}\n\n", std::current_exception());
```

but i understand, probably we should change all of them in another change, as we are writing error messages to `stdout` in app template already.

xemul replied:
Yes, that's separate patch. Here I made all error logging consistent

tchaikov commented on include/seastar/util/program-options.hh:
```diff
@@ -456,6 +471,61 @@ public:
+
+        T val;
+        if (mutator.visit_value(val)) {
+            value<T>::do_set_value(std::move(val), false);
```
nit, to avoid repeating the dependent name of the parent class, probably could just use `this`, like
```c++
this->do_set_value(std::move(val), false);
```
less repeating this way.

xemul commented:
upd (applied @tchaikov  suggestions around tilde expansion):
- check config_file not to be empty
- abort on unset HOME early and explicitly
- .replace() tilde with raw home, w/o converting it to std::string

xemul commented:
@denesb , merge ping

denesb commented:
> @denesb , merge ping

I'm not a seastar maintainer.

xemul commented:
@avikivity , review ping

xemul commented:
@avikivity , review ping

xemul commented:
@avikivity , review ping

xemul commented:
@avikivity , review ping

xemul commented:
upd:
- rebased
- added sugar for reactor options cross-shard updating
- made more options live-updateable

@avikivity , your review is very much welcome

xemul commented:
@avikivity , please review

avikivity commented:
Having Seastar be responsible for reading options was a design mistake IMO. The app should be responsible for options, and of it needs to send some to Seastar that it should do so via app_template::config. Then, the names of the parameters, the format and location of the configuration file, and responsibility for reloading are all up to the application.

So, we need to choose. Do we continue to improve option processing, or do we deprecate them in favor of moving everything to the application?

xemul commented:
I partially agree with that and the "it needs to send some to Seastar that it should do so via app_template::config" was mostly addressed by the 8810950b92eaa7f4115eb19e3294a5c6c1af8cb9 work. The only difference is that the configuration is fed via `app_template::seastar_options`, not `app_template::config`. And this PR mostly touches the `program_options` to make it live-updateable. Hunks to inject HUP handling and the `--config-file` option are there just because that's the current way `program_options` are populated.

As far as "the names of the parameters, the format and location of the configuration file, and responsibility for reloading are all up to the application" is concerned. Let's take for example fbb9fcd15a7e15c24b75f85b003723bea29d85e5. If this option was added without seastar support for the option CLI name, then all apps using seastar would need to add the option on their own, which's not very convenient.  I think that it's still worth keeping the config and/or CLI parsing in seastar, but maybe isolating it even more in util so that it's fully optional.

avikivity commented:
I though to keep exposing bpo helpers, so a bpo application can continue ask Seastar to add Seastar options to the application bpo parser.

avikivity commented on src/core/app-template.cc:
```diff
@@ -118,11 +119,25 @@ const app_template::seastar_options& app_template::options() const {
+        if (config_file.empty()) {
+            throw std::runtime_error("Empty 'config-file'");
+        }
+        if (config_file[0] == '~') {
```
~ expands to $HOME only if not followed by a word. ~user = user's home.

avikivity replied:
Is it needed here? Let bash perform the expansion, and if it's not run from a bash script, then the user should be explict.

avikivity replied:
Won't it just work? bash will expand it.

```sh
$ /bin/echo ~/foo
/home/avi/foo
```

xemul replied:
Yes, provided it's run from bash without quotes
```
$ /bin/echo '~/'
~/
```
and without fancy variabling
```
$ FOO='~/'; echo ${FOO}
~/
```

But it's indeed doesn't worth the code to parse. I'll drop it

avikivity commented on src/core/reactor.cc:
```diff
@@ -3776,7 +3776,9 @@ reactor_options::reactor_options(program_options::option_group* parent_group)
-    , unsafe_bypass_fsync(*this, "unsafe-bypass-fsync", false, "Bypass fsync(), may result in data loss. Use for testing on consumer drives")
+    , unsafe_bypass_fsync(*this, "unsafe-bypass-fsync", false, "Bypass fsync(), may result in data loss. Use for testing on consumer drives",
+            update_option_on_all_shards<bool>([] (auto v) { engine().set_bypass_fsync(v); })
+    )
```
unsafe-bypass-fsync is for developers, I don't see a real case for changing it at runtime.

xemul replied:
It's just for demonstration how it would look like

avikivity commented:
I think this is problematic, it adds a new configuration file format for the application. For ScyllaDB, it means an extra source of configuration outside scylla.yaml.

avikivity commented:
> > I think this is problematic, it adds a new configuration file format for the application. For ScyllaDB, it means an extra source of configuration outside scylla.yaml.
> 
> One cannot tune seastar options via `scylla.yaml` even now. I do not remember the exact file names and options, but if someone needs to add seastar options to installed-and-running scylla, one needs to find SCYLLA_OPTIONS (?) variable in /etc//scylla. shell script (with variables) and put it _there_ as a CLI option, not as .yaml/.json/.ini entry. So it's pain anyway -- either there are two config files, or the need to teach scylla.yaml to configure seastar options.

Sure, but isn't it overall nicer to have options in one place? I'm thinking of --blocked-reactor-notify-ms (which might as well be renamed).

avikivity commented:
btw, it's a little problematic since the file has to be accessed both via Seastar APIs and via standard APIs (before launch), but that's life.
